### PR TITLE
feat: React 16 API Implementation

### LINKED ISSUE #147: HMR stuck in infinite loop
I've noticed that HMR is broken for me because the universal HOC runs into an infinite loop on `componentWillReceiveProps`. Specifically removing this state update fixes it for me: https://github.com/faceyspacey/react-universal-component/blob/master/src/index.js#L196.

Unless I am missing something, what is there to guard against multiple hits to the lifecycle method with that state update (leading to an infinite re-render cycle)?

### PR DESCRIPTION
**Pull request by re6exp**
- Changed `react` and `react-test-renderer` versions;
- Moved from using `componentWillMount` to constructor, `componentWillReceiveProps` to `getDerivedStateFromProps`.

**NOTICE**: description page should be changed, I didn't do this.

Resolves #147 #129

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -192,23 +214,34 @@ export default function universal<Props: Props>(
...
+    }
+
+    requireAsyncInner(
+      requireAsync: RequireAsync,
+      props: Props,
+      state: State,
+      context: Context,
+      isMount?: boolean
+    ) {
+      if (state.mod && loadingTransition) {
```
Comment (ScriptedAlchemy) [inline src/index.js]: Change to:
```    if (!state.mod && loadingTransition) {
        this.update({ mod: null, props }) // display `loading` during componentWillReceiveProps
      }```

and Loading state works

**Comment by re6exp** [comment]
Great!

вс, 9 сент. 2018 г., 20:29 Zack Jackson <notifications@github.com>:

File: src/index.js (hunk 1 above)
Comment (re6exp) [reply src/index.js]: Yep...

**Comment by ScriptedAlchemy** [comment]
Good cleanup! Ill release a next tag on npm for this. Ill test this further as well

**Comment by ScriptedAlchemy** [comment]
Merging this into my branch for easier management. open any extra react 16 stuff to v5 branch

**Comment by ScriptedAlchemy** [comment]
npm tag `@react16`

**Pull request merged**